### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by Zubayer204**
I have edited the code a bit so that it is compatible with python 3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,89 +0,0 @@
-# Sega Genesis Checksum Utility
```
Comment (mrhappyasthma) [inline README.md]: Why is this file deleted? Can we revert this?

File: Resources/Backup/The Complete Documentation About Genesis Rom Format (v1.1).txt [hunk 2]
```diff
@@ -1,375 +0,0 @@
-// NOTE: Backed up copy from http://www.emulatronia.com/doctec/consolas/megadrive/genesis_rom.txt.
```
Comment (mrhappyasthma) [inline Resources/Backup/The Complete Documentation About Genesis Rom Format (v1.1).txt]: Why is this file deleted? Can we revert this?

File: sega_genesis_checksum_utility_for_python_3.py [hunk 3]
```diff
@@ -25,46 +25,46 @@ def main():
                                             'Genesis ROM binary.')
   args = parser.parse_args()
 
-  print 'Reading checksum from file...'
-  with open(args.path[0], 'r+b') as genesis_file:
+  print('Reading checksum from file...')
+  with open(args.path[0], 'rb') as genesis_file:
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: Doesn't this need to be `r+b` for the `write_checksum_to_header`?

Otherwise it's only open for reading.

File: sega_genesis_checksum_utility_for_python_3.py [hunk 4]
```diff
@@ -25,46 +25,46 @@ def main():
...
-      print '\nChecksums match. Horray!'
+      print('\nChecksums match. Horray!')
       return
 
-    print '\nWARNING: Checksums do not match!'
+    print('\nWARNING: Checksums do not match!')
     while(1):
-      data = raw_input('\nWould you like to update the header checksum to match'
+      data = input('\nWould you like to update the header checksum to match'
                        'match the computed checksum? (y/n)')
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: nitpick: Fix alignment of the strings for `input()`

File: sega_genesis_checksum_utility_for_python_3.py [hunk 5]
```diff
@@ -142,12 +142,12 @@ def verify_console_name_from_header(open_file):
     True if the header of `open_file` contains the correct console name in the
     right location.
   """
-  CONSOLE_NAME_OFFSET = 0x100
-  open_file.seek(CONSOLE_NAME_OFFSET)
-  console_name = open_file.read(15)
+  memorybuffer = open_file.read()
+  memorybuffer[0x100:0x10c].decode('utf-8')
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: Why are we calling `memorybuffer[0x100:0x10c].decode('utf-8')` twice? Once on line 146 and once on line 147. Can the line 146 just be removed entirely?

File: sega_genesis_checksum_utility_for_python_3.py [hunk 6]
```diff
@@ -142,12 +142,12 @@ def verify_console_name_from_header(open_file):
...
+  console_name = memorybuffer[0x100:0x10c].decode('utf-8')
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: Doesn't this only decode 12 characters?

SEGA MEGA DRIVE would be 15.

File: sega_genesis_checksum_utility_for_python_3.py [hunk 7]
```diff
@@ -142,12 +142,12 @@ def verify_console_name_from_header(open_file):
...
   if console_name == "SEGA MEGA DRIVE":
     return True
-  if console_name[:12] == "SEGA GENESIS":
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: Conceptual question: did any of this logic need to change? Isn't it still functional in python3?

**Review by mrhappyasthma — CHANGES_REQUESTED** [review]
Generally the changes look pretty good, but put a few comments for clarification or suggested improvements.

File: README.md (hunk 1 above)
Comment (Zubayer204) [reply README.md]: > Why is this file deleted? Can we revert this?
Yes, of course. I am adding it to my pull request.

Comment (mrhappyasthma) [reply README.md]: It looks like the pull request still tries to delete the README. Can you upload that one as well?

Once you do that and address the comments below, I'll be happy to merge this in.

File: sega_genesis_checksum_utility_for_python_3.py (hunk 3 above)
Comment (Zubayer204) [reply sega_genesis_checksum_utility_for_python_3.py]: Yes, it caught it. It should be r+b to run on any os. I am changing it. But it will still work on windows if it is rb

File: Resources/Backup/The Complete Documentation About Genesis Rom Format (v1.1).txt (hunk 2 above)
Comment (Zubayer204) [reply Resources/Backup/The Complete Documentation About Genesis Rom Format (v1.1).txt]: Reverted it.

File: sega_genesis_checksum_utility_for_python_3.py (hunk 4 above)
Comment (Zubayer204) [reply sega_genesis_checksum_utility_for_python_3.py]: Didn't understood properly about what you mean. But in python3 the raw_input() function is not available. It is merged into input()

File: sega_genesis_checksum_utility_for_python_3.py (hunk 3 above)
Comment (mrhappyasthma) [reply sega_genesis_checksum_utility_for_python_3.py:29]: Sorry for not being more clear.

This is not a problem with the code itself, but just how it's formatted.

Can we remove the spaces on the wrapped line so that the quotes of the text are aligned?

e.g. https://i.imgur.com/xjqEBTn.jpg

File: sega_genesis_checksum_utility_for_python_3.py (hunk 5 above)
Comment (Zubayer204) [reply sega_genesis_checksum_utility_for_python_3.py]: Yeah, removed it entirely.

File: sega_genesis_checksum_utility_for_python_3.py (hunk 6 above)
Comment (Zubayer204) [reply sega_genesis_checksum_utility_for_python_3.py]: Yeah, my fault. changed it.

File: sega_genesis_checksum_utility_for_python_3.py (hunk 7 above)
Comment (Zubayer204) [reply sega_genesis_checksum_utility_for_python_3.py]: I am removing my current logic and using another logic. `if console_name.strip() == ""SEGA GENESIS"" `  I am using strip method to remove the whitespace before and after the console name. It is REQUIRED because the console name might contain whitespace and don't execute the logic.

File: README.md (hunk 1 above)
Comment (Zubayer204) [reply README.md]: > It looks like the pull request still tries to delete the README. Can you upload that one as well?
> 
> Once you do that and address the comments below, I'll be happy to merge this in.

Did all of that.

File: sega_genesis_checksum_utility_for_python_3.py (hunk 4 above)
Comment (mrhappyasthma) [reply sega_genesis_checksum_utility_for_python_3.py]: Sorry for not being more clear.

This is not a problem with the code itself, but just how it's formatted.

Can we remove the spaces on the wrapped line so that the quotes of the text are aligned?

e.g. https://i.imgur.com/xjqEBTn.jpg

File: sega_genesis_checksum_utility_for_python_3.py (hunk 7 above)
Comment (mrhappyasthma) [reply sega_genesis_checksum_utility_for_python_3.py]: Seems reasonable. Tested it out locally and seems to work fine.

**Comment by mrhappyasthma** [comment]
I just noticed, this adds a new file with the `_for_python_3` suffix.

Since python2 is no longer being supported, can we just update the existing file. I don't see any benefits in maintaining both versions. (In other words, rename your new file back to the original file name `sega_genesis_checksum_utility.py`). This will also update the pull request diff to show only the updated lines.

File: sega_genesis_checksum_utility_for_python_3.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python
```
Comment (mrhappyasthma) [inline sega_genesis_checksum_utility_for_python_3.py]: I just noticed, this adds a new file with the `_for_python_3` suffix.

Since python2 is no longer being supported, can we just update the existing file. I don't see any benefits in maintaining both versions. (In other words, rename your new file back to the original file name `sega_genesis_checksum_utility.py`). This will also update the pull request diff to show only the updated lines.

**Comment by mrhappyasthma** [comment]
I would prefer we revert most of the comments that you added. The majority of them are self-documented by the code.

https://en.wikipedia.org/wiki/Self-documenting_code

For instance:

```
   if header_checksum == computed_checksum: #if the computed checksum and retrived checksum are same then it is a valid file
      print '\nChecksums match. Horray!')
```

I don't think this comment is needed. It's clear that we are checking if two checksums are equal based on the first line . And then it has a print statement confirming the match. The comment doesn't provide any additional clarity.

The only one that I think adds more context than the code itself describes is:

```
console_name = memorybuffer[0x100:0x110].decode('utf-8') #get the 15bits console_name and decode it with unicode
```

But I think instead of leaving a comment we can make the code a bit more self-describing. Perhaps by using local variable constants.

Maybe we could do something like:

```
CONSOLE_NAME_OFFSET = 0x100
CONSOLE_NAME_MAX_SIZE = 15

console_name = memorybuffer[CONSOLE_NAME_OFFSET:(CONSOLE_NAME_OFFSET + CONSOLE_NAME_MAX_SIZE)].decode('utf-8')
```

This way the local variable constants have meaningful names to describe the magic numbers (the header offset of `0x100` and the max size of `15`).

And the value we are decoding from the memory buffer is stored in `console_name`. So then the code itself becomes more clear (without the need for any comment) as to what we are reading from memory and storing in the `console_name`.

**Comment by mrhappyasthma** [comment]
Any update on this?

**Comment by Zubayer204** [comment]
> Any update on this?

Sorry, I am freelancer. I didn't get much time to work on this.

**Comment by mrhappyasthma** [comment]
No worries. Take your time. Just wanted to check that this was still on your radar.

**Comment by mrhappyasthma** [comment]
I'm also using python 3 now, so merging this in.

I forgot to squash it myself, so I did so using a rebase in https://github.com/mrhappyasthma/Sega-Genesis-Checksum-Utility/commit/1add81ee34f8dabd8abe4f5080c5aa1553e1b8c9
